fdbrpc/include/fdbrpc/FileTransfer.h

```diff
@@ -0,0 +1,263 @@
+
+	std::optional<GetFileInfoReply> GetFileInfo(const std::string& filename, bool get_crc_checksum = false) {
+		ClientContext context;
+		GetFileInfoRequest request;
+		request.set_file_name(filename);
+		request.set_get_size(true);
+		request.set_get_crc_checksum(get_crc_checksum);
+
+		GetFileInfoReply response;
+		Status status = stub_->GetFileInfo(&context, request, &response);
+		if (status.ok()) {
+			return { response };
+		} else {
+			return std::nullopt;
+		}
+	}
+
+	// Downloads a file from a remote server and saves it locally.
+	//
+	// This function communicates with a remote server to download a file specified by `filename`
+	// and saves it to the local file system with the name `output_filename`. Optionally, it can
+	// verify the integrity of the downloaded file using a CRC checksum.
+	//
+	// Params:
+	// - filename: The name of the file to download from the remote server.
+	// - output_filename: The name of the file to save the downloaded content locally.
+	// - verify: Whether to verify the file's integrity using CRC checksum
+	//
+	// Returns:
+	//  std::optional<size_t> The size of the downloaded file in bytes if successful, or std::nullopt if the download
+	//  failed. `output_filename` is deleted on failure.
+	std::optional<size_t> DownloadFile(const std::string& filename,
+	                                   const std::string& output_filename,
+	                                   bool verify = true) {
+
+		uint32_t expected_crc = 0;
+		uint32_t expected_size = 0;
+		{
+			GetFileInfoRequest request;
+			ClientContext context;
+			request.set_file_name(filename);
+			request.set_get_crc_checksum(verify);
+			request.set_get_size(true);
+			GetFileInfoReply response;
+			auto res = stub_->GetFileInfo(&context, request, &response);
+			if (!res.ok()) {
+				return std::nullopt;
+			}
+			expected_crc = response.crc_checksum();
+			expected_size = response.file_size();
+		}
+
+		DownloadRequest request;
+		request.set_file_name(filename);
+
+		ClientContext context;
+		std::unique_ptr<grpc::ClientReader<DownloadChunk>> reader(stub_->DownloadFile(&context, request));
+
+		std::ofstream output_file(output_filename, std::ios::binary | std::ios::trunc);
+		if (!output_file.is_open()) {
+			return std::nullopt; // Failed to open file
+		}
+
+		DownloadChunk chunk;
+		size_t bytes_read = 0;
+		bool failed = false;
+		while (reader->Read(&chunk)) {
+			if (chunk.offset() != bytes_read) {
+				// Abort on invalid offset
+				failed = true;
+				break;
+			}
+
+			output_file.write(chunk.data().data(), chunk.data().size());
+			bytes_read += chunk.data().size();
+		}
+
+		// Close file after writing
+		output_file.close();
+		failed = failed || (bytes_read != expected_size);
+
+		// Verify checksum
+		if (!failed && verify) {
+			std::ifstream output_file_reader(output_filename);
+			uint32_t actual_crc = crc32_checksum_ifstream(&output_file_reader);
+			failed = (actual_crc != expected_crc);
```

Comment:
How do I distingush between size mismatch, crc mismatch or fail for another reason?

Reply:
Log?

Reply:
It a TODO in the next comment you made :), but yes I can also do logging. But since this is fdbrpc module I don't have access to TraceEvent here.

Reply:
Actually may be wrong about TraceEvent. I see LoadBalance having one.

Reply:
ack